src/coreclr/jit/llvm.cpp

```diff
@@ -1083,7 +1083,25 @@ llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
     for (GenTreeCall::Use& use : call->Args())
     {
         lastArg = use.GetNode()->AsPutArgType();
-        argVec.push_back(consumeValue(lastArg->gtGetOp1(), getLlvmTypeForCorInfoType(lastArg->GetCorInfoType(), lastArg->GetClsHnd())));
+
+        if (lastArg->gtGetOp1()->OperIs(GT_FIELD_LIST))
+        {
+            assert(lastArg->GetCorInfoType() == CorInfoType::CORINFO_TYPE_VALUECLASS);
+
+            // alloca a type using the lastArg type
+            Type* argLlvmType = getLlvmTypeForCorInfoType(lastArg->GetCorInfoType(), lastArg->GetClsHnd());
+            Value* allocaValue = _builder.CreateAlloca(argLlvmType);
+
+            // TODO-LLVM: store LLVM struct type or GetClsHnd somewhere to facilitate compiling the GT_FIELD_LIST
```

Comment:
This could look something like this:
1) `buildFieldList` should take the LLVM type of the struct and return the loaded `Value*`.
2) Here we would do:
```c
Type*    argLlvmType = getLlvmTypeForCorInfoType(lastArg->GetCorInfoType(), lastArg->GetClsHnd());
GenTree* argNode     = lastArg->gtGetOp1();
Value*   argValue;
if (lastArg->gtGetOp1()->OperIs(GT_FIELD_LIST))
{
    argValue = buildFieldList(argNode->AsFieldList());
}
else
{
    argValue = consumeValue(argNode, argType);
}

argVec.push_back(argValue);
```
It's a little unsatisfying to not build the field list when encountering it in the linear order, but the alternative is to store the arg's type handle in lowering, which isn't meaningfully better.

Reply:
Thanks, do you anticipate that I'll need to do the same for `STORE_OBJ`, i.e. for `t88`/`155` here?
```
------------ BB01 [000..021) -> BB03 (cond), preds={} succs={BB02,BB03}
               [000100] ------------       t100 =    LCL_VAR   int    V08 rat0         u:1
               [000099] ------------        t99 =    CNS_INT   int    0
                                                  /--*  t100   int
                                                  +--*  t99    int
               [000101] ------------       t101 = *  ADD       int
                                                  /--*  t101   int
N001 (  1,  1) [000002] ------------         t2 = *  IND       ref
N002 (  1,  1) [000064] ------------        t64 =    CNS_INT   int    4 field offset Fseq[_callback] $42
                                                  /--*  t2     ref
                                                  +--*  t64    int
N003 (  2,  2) [000065] -------N----        t65 = *  ADD       byref  $c0
                                                  /--*  t65    byref
N004 (  4,  4) [000003] ---XG-------         t3 = *  IND       ref    <l:$144, c:$143>
               [000103] ------------       t103 =    LCL_VAR   int    V08 rat0         u:1
               [000102] ------------       t102 =    CNS_INT   int    8
                                                  /--*  t103   int
                                                  +--*  t102   int
               [000104] ------------       t104 = *  ADD       int
                                                  /--*  t104   int
                                                  +--*  t3     ref
N006 (  4,  4) [000005] -A-XG-------              *  STOREIND  ref
               [000106] ------------       t106 =    LCL_VAR   int    V08 rat0         u:1
               [000105] ------------       t105 =    CNS_INT   int    0
                                                  /--*  t106   int
                                                  +--*  t105   int
               [000107] ------------       t107 = *  ADD       int
                                                  /--*  t107   int
N001 (  1,  1) [000006] ------------         t6 = *  IND       ref
N002 (  1,  1) [000066] ------------        t66 =    CNS_INT   int    4 field offset Fseq[_callback] $42
                                                  /--*  t6     ref
                                                  +--*  t66    int
N003 (  2,  2) [000067] -------N----        t67 = *  ADD       byref  $c0
N005 (  1,  1) [000007] ------------         t7 =    CNS_INT   ref    null $VN.Null
                                                  /--*  t67    byref
                                                  +--*  t7     ref
               [000098] -A--GO------              *  STOREIND  ref
               [000109] ------------       t109 =    LCL_VAR   int    V08 rat0         u:1
               [000108] ------------       t108 =    CNS_INT   int    0
                                                  /--*  t109   int
                                                  +--*  t108   int
               [000110] ------------       t110 = *  ADD       int
                                                  /--*  t110   int
N001 (  1,  1) [000011] ------------        t11 = *  IND       ref
N002 (  1,  1) [000068] ------------        t68 =    CNS_INT   int    8 field offset Fseq[_state] $43
                                                  /--*  t11    ref
                                                  +--*  t68    int
N003 (  3,  3) [000069] ------------        t69 = *  ADD       byref  $c1
               [000112] ------------       t112 =    LCL_VAR   int    V08 rat0         u:1
               [000111] ------------       t111 =    CNS_INT   int    4
                                                  /--*  t112   int
                                                  +--*  t111   int
               [000113] ------------       t113 = *  ADD       int
                                                  /--*  t113   int
                                                  +--*  t69    byref
N005 (  3,  3) [000071] -A----------              *  STOREIND  byref
               [000115] ------------       t115 =    LCL_VAR   int    V08 rat0         u:1
               [000114] ------------       t114 =    CNS_INT   int    4
                                                  /--*  t115   int
                                                  +--*  t114   int
               [000116] ------------       t116 = *  ADD       int
                                                  /--*  t116   int
N006 (  1,  1) [000073] ------------        t73 = *  IND       byref  Zero Fseq[Item1]
                                                  /--*  t73    byref
N007 (  3,  2) [000074] n----O------        t74 = *  IND       ref    <l:$146, c:$145>
               [000118] ------------       t118 =    LCL_VAR   int    V08 rat0         u:1
               [000117] ------------       t117 =    CNS_INT   int    16
                                                  /--*  t118   int
                                                  +--*  t117   int
               [000119] ------------       t119 = *  ADD       int
                                                  /--*  t119   int
                                                  +--*  t74    ref
N009 (  7,  5) [000075] -A---O------              *  STOREIND  ref
               [000121] ------------       t121 =    LCL_VAR   int    V08 rat0         u:1
               [000120] ------------       t120 =    CNS_INT   int    4
                                                  /--*  t121   int
                                                  +--*  t120   int
               [000122] ------------       t122 = *  ADD       int
                                                  /--*  t122   int
N011 (  1,  1) [000078] ------------        t78 = *  IND       byref
N012 (  1,  1) [000079] ------------        t79 =    CNS_INT   int    4 Fseq[Item2] $42
                                                  /--*  t78    byref
                                                  +--*  t79    int
N013 (  2,  2) [000080] -------N----        t80 = *  ADD       byref  $c2
                                                  /--*  t80    byref
N014 (  4,  4) [000081] n----O------        t81 = *  IND       ref    <l:$148, c:$147>
               [000124] ------------       t124 =    LCL_VAR   int    V08 rat0         u:1
               [000123] ------------       t123 =    CNS_INT   int    20
                                                  /--*  t124   int
                                                  +--*  t123   int
               [000125] ------------       t125 = *  ADD       int
                                                  /--*  t125   int
                                                  +--*  t81    ref
N016 (  8,  7) [000082] -A---O------              *  STOREIND  ref
               [000127] ------------       t127 =    LCL_VAR   int    V08 rat0         u:1
               [000126] ------------       t126 =    CNS_INT   int    8
                                                  /--*  t127   int
                                                  +--*  t126   int
               [000128] ------------       t128 = *  ADD       int
                                                  /--*  t128   int
N001 (  1,  1) [000010] ------------        t10 = *  IND       ref
N002 (  1,  1) [000084] ------------        t84 =    CNS_INT   int    4 field offset Fseq[m_firstParameter] $42
                                                  /--*  t10    ref
                                                  +--*  t84    int
N003 (  2,  2) [000085] -------N----        t85 = *  ADD       byref  <l:$c4, c:$c3>
                                                  /--*  t85    byref
N004 (  4,  4) [000018] ---XG-------        t18 = *  IND       ref    <l:$14e, c:$14d>
               [000130] ------------       t130 =    LCL_VAR   int    V08 rat0         u:1
               [000129] ------------       t129 =    CNS_INT   int    12
                                                  /--*  t130   int
                                                  +--*  t129   int
               [000131] ------------       t131 = *  ADD       int
                                                  /--*  t131   int
                                                  +--*  t18    ref
N006 (  4,  4) [000025] -A-XG-------              *  STOREIND  ref
               [000133] ------------       t133 =    LCL_VAR   int    V08 rat0         u:1
               [000132] ------------       t132 =    CNS_INT   int    8
                                                  /--*  t133   int
                                                  +--*  t132   int
               [000134] ------------       t134 = *  ADD       int
                                                  /--*  t134   int
N001 (  1,  1) [000020] ------------        t20 = *  IND       ref
N002 (  1,  1) [000086] ------------        t86 =    CNS_INT   int    16 field offset Fseq[m_functionPointer] $45
                                                  /--*  t20    ref
                                                  +--*  t86    int
N003 (  2,  2) [000087] -------N----        t87 = *  ADD       byref  <l:$c6, c:$c5>
                                                  /--*  t87    byref
N004 (  4,  4) [000021] n---GO------        t21 = *  IND       int    <l:$243, c:$242>
                                                  /--*  t21    int
N006 (  4,  4) [000023] DA--GO------              *  STORE_LCL_VAR int    V03 tmp1         d:1
N001 (  1,  1) [000032] ------------        t32 =    LCL_VAR   int    V03 tmp1         u:1 <l:$241, c:$280>
N002 (  1,  4) [000031] ------------        t31 =    CNS_INT   int    -0x80000000 $46
                                                  /--*  t32    int
                                                  +--*  t31    int
N003 (  3,  6) [000033] ------------        t33 = *  AND       int    <l:$245, c:$244>
N004 (  1,  1) [000034] ------------        t34 =    CNS_INT   int    0 $40
                                                  /--*  t33    int
                                                  +--*  t34    int
N005 (  5,  8) [000035] J------N----        t35 = *  NE        int    <l:$247, c:$246>
                                                  /--*  t35    int
N006 (  7, 10) [000036] ------------              *  JTRUE     void

------------ BB02 [000..000) -> BB04 (always), preds={BB01} succs={BB04}
               [000136] ------------       t136 =    LCL_VAR   int    V08 rat0         u:1
               [000135] ------------       t135 =    CNS_INT   int    16
                                                  /--*  t136   int
                                                  +--*  t135   int
               [000137] ------------       t137 = *  ADD       int
                                                  /--*  t137   int
N002 (  3,  2) [000041] -------N----        t41 = *  IND       ref
               [000139] ------------       t139 =    LCL_VAR   int    V08 rat0         u:1
               [000138] ------------       t138 =    CNS_INT   int    20
                                                  /--*  t139   int
                                                  +--*  t138   int
               [000140] ------------       t140 = *  ADD       int
                                                  /--*  t140   int
N003 (  3,  2) [000089] ------------        t89 = *  IND       ref
                                                  /--*  t41    ref
                                                  +--*  t89    ref
N004 (  6,  4) [000088] -c----------        t88 = *  FIELD_LIST struct $301
               [000142] ------------       t142 =    LCL_VAR   int    V08 rat0         u:1
               [000141] ------------       t141 =    CNS_INT   int    12
                                                  /--*  t142   int
                                                  +--*  t141   int
               [000143] ------------       t143 = *  ADD       int
                                                  /--*  t143   int
N005 (  1,  1) [000038] ------------        t38 = *  IND       ref
N006 (  1,  1) [000042] ------------        t42 =    LCL_VAR   int    V03 tmp1         u:1 (last use) <l:$241, c:$280>
               [000144] ------------       t144 =    LCL_VAR   int    V08 rat0         u:1
               [000145] ------------       t145 =    CNS_INT   int    32
                                                  /--*  t144   int
                                                  +--*  t145   int
               [000146] ------------       t146 = *  ADD       int
                                                  /--*  t146   int
               [000147] ------------       t147 = *  PUTARG_TYPE int
               [000148] ------------       t148 =    LCL_VAR   int    V08 rat0         u:1
               [000149] ------------       t149 =    CNS_INT   int    32
                                                  /--*  t148   int
                                                  +--*  t149   int
               [000150] ------------       t150 = *  ADD       int
                                                  /--*  t150   int
                                                  +--*  t38    ref
               [000151] ------------              *  STOREIND  ref
               [000152] ------------       t152 =    LCL_VAR   int    V08 rat0         u:1 (last use)
               [000153] ------------       t153 =    CNS_INT   int    36
                                                  /--*  t152   int
                                                  +--*  t153   int
               [000154] ------------       t154 = *  ADD       int
                                                  /--*  t154   int
                                                  +--*  t88    struct
               [000155] s---G-------              *  STORE_OBJ struct<System.ValueTuple`2[System.__Canon, System.__Canon], 8> (copy)
                                                  /--*  t147   int    arg-1
                                                  +--*  t42    int    calli tgt
N007 ( 28, 11) [000037] --CXG-------              *  CALL ind  void   $VN.Void

------------ BB03 [000..000), preds={BB01} succs={BB04}
N002 (  1,  1) [000048] ------------        t48 =    LCL_VAR   int    V03 tmp1         u:1 <l:$241, c:$280>
N003 (  1,  4) [000049] ------------        t49 =    CNS_INT   int    -0x7FFFFFFC $47
                                                  /--*  t48    int
                                                  +--*  t49    int
N004 (  2,  5) [000047] -------N----        t47 = *  ADD       int    <l:$249, c:$248>
                                                  /--*  t47    int
N005 (  4,  7) [000052] ---X--------        t52 = *  IND       int    <l:$24b, c:$24a>
                                                  /--*  t52    int
N006 (  7,  9) [000053] ---X--------        t53 = *  IND       int    <l:$24f, c:$24e>
               [000157] ------------       t157 =    LCL_VAR   int    V08 rat0         u:1
               [000156] ------------       t156 =    CNS_INT   int    16
                                                  /--*  t157   int
                                                  +--*  t156   int
               [000158] ------------       t158 = *  ADD       int
                                                  /--*  t158   int
N007 (  3,  2) [000058] -------N----        t58 = *  IND       ref
               [000160] ------------       t160 =    LCL_VAR   int    V08 rat0         u:1
               [000159] ------------       t159 =    CNS_INT   int    20
                                                  /--*  t160   int
                                                  +--*  t159   int
               [000161] ------------       t161 = *  ADD       int
                                                  /--*  t161   int
N008 (  3,  2) [000095] ------------        t95 = *  IND       ref
                                                  /--*  t58    ref
                                                  +--*  t95    ref
N009 (  6,  4) [000094] -c----------        t94 = *  FIELD_LIST struct $300
               [000163] ------------       t163 =    LCL_VAR   int    V08 rat0         u:1
               [000162] ------------       t162 =    CNS_INT   int    12
                                                  /--*  t163   int
                                                  +--*  t162   int
               [000164] ------------       t164 = *  ADD       int
                                                  /--*  t164   int
N010 (  1,  1) [000055] ------------        t55 = *  IND       ref
N011 (  1,  1) [000043] ------------        t43 =    LCL_VAR   int    V03 tmp1         u:1 (last use) <l:$241, c:$280>
N012 (  1,  4) [000044] ------------        t44 =    CNS_INT   int    0x80000000 $46
                                                  /--*  t43    int
                                                  +--*  t44    int
N013 (  2,  5) [000045] -------N----        t45 = *  ADD       int    <l:$251, c:$250>
                                                  /--*  t45    int
N014 (  5,  7) [000046] ---X--------        t46 = *  IND       int    <l:$253, c:$252>
               [000165] ------------       t165 =    LCL_VAR   int    V08 rat0         u:1
               [000166] ------------       t166 =    CNS_INT   int    32
                                                  /--*  t165   int
                                                  +--*  t166   int
               [000167] ------------       t167 = *  ADD       int
                                                  /--*  t167   int
               [000168] ------------       t168 = *  PUTARG_TYPE int
               [000169] ------------       t169 =    LCL_VAR   int    V08 rat0         u:1
               [000170] ------------       t170 =    CNS_INT   int    32
                                                  /--*  t169   int
                                                  +--*  t170   int
               [000171] ------------       t171 = *  ADD       int
                                                  /--*  t171   int
                                                  +--*  t55    ref
               [000172] ------------              *  STOREIND  ref
                                                  /--*  t53    int
               [000173] ---X--------       t173 = *  PUTARG_TYPE int
               [000174] ------------       t174 =    LCL_VAR   int    V08 rat0         u:1 (last use)
               [000175] ------------       t175 =    CNS_INT   int    36
                                                  /--*  t174   int
                                                  +--*  t175   int
               [000176] ------------       t176 = *  ADD       int
                                                  /--*  t176   int
                                                  +--*  t94    struct
               [000177] s---G-------              *  STORE_OBJ struct<System.ValueTuple`2[System.__Canon, System.__Canon], 8> (copy)
                                                  /--*  t168   int    arg-1
                                                  +--*  t173   int    arg1
                                                  +--*  t46    int    calli tgt
N015 ( 42, 27) [000054] --CXG-------              *  CALL ind  void   $VN.Void

------------ BB04 [000..000) (return), preds={BB02,BB03} succs={}
N001 (  0,  0) [000016] ------------                 RETURN    void   $340

-------------------------------------------------------------------------------------------------------------------
*************** In fgDebugCheckBBlist

*************** Starting PHASE Build LLVM
```

Reply:
ah right, let me see why that didn't happen.

Reply:
(It would have to supported explicitly, I don't think code for it exists already)

Reply:
oh, I've not done it yet, it's the comment later on I think. lol.

Reply:
or not, that's for parameters.  Do you want this code in this PR, guess I might as well for completeness?

Reply:
No strong opinion, up to you.

Reply:
> I would expect that case to be handled in lowering by eliminating the `FIELD_LIST` and storing individual fields directly to the shadow stack.

I've had a stab at this in https://github.com/dotnet/runtimelab/pull/2007/commits/bd8e19f10b64221a397bef69e6eaeafe3c2ce6b7